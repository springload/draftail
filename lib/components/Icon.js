import PropTypes from "prop-types";
import React from "react";

/**
 * Icon as SVG element. Can optionally render a React element instead.
 */
const Icon = ({ icon, title, className }) => {
  const isPathOrRef = typeof icon === "string";
  let children;

  if (isPathOrRef) {
    if (icon.includes("#")) {
      children = <use xlinkHref={icon} />;
    } else {
      children = <path d={icon} />;
    }
  } else if (Array.isArray(icon)) {
    // eslint-disable-next-line @thibaudcolas/cookbook/react/no-array-index-key
    children = icon.map((d, i) => <path key={i} d={d} />);
  } else {
    return icon;
  }

  return (
    <svg
      width="16"
      height="16"
      viewBox="0 0 1024 1024"
      className={`Draftail-Icon ${className || ""}`}
      aria-hidden={title ? null : true}
      role={title ? "img" : null}
      aria-label={title || null}
    >
      {children}
    </svg>
  );
};

Icon.propTypes = {
  // The icon definition is very flexible.
  icon: PropTypes.oneOfType([
    // String icon = SVG path or symbol reference.
    PropTypes.string,
    // List of SVG paths.
    PropTypes.arrayOf(PropTypes.string),
    // Arbitrary React element.
    PropTypes.node,
  ]).isRequired,
  title: PropTypes.string,
  className: PropTypes.string,
};

Icon.defaultProps = {
  title: null,
  className: null,
};

export default Icon;
