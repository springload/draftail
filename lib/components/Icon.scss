.#{$NS}Icon {
    fill: currentColor;
    pointer-events: none;
}
