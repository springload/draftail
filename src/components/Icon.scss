@use "../api/constants" as *;

.#{$DRAFTAIL}Icon {
  fill: currentColor;
  pointer-events: none;
  vertical-align: middle;
}
