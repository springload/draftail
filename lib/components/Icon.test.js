import PropTypes from 'prop-types';
import React from 'react';
import { shallow } from 'enzyme';
import Icon from '../components/Icon';

const SQUARE = 'M10 10 H 90 V 90 H 10 Z';

const CustomIcon = ({ icon }) => (
    <span className={`icon icon-${icon}`} aria-hidden />
);

CustomIcon.propTypes = {
    icon: PropTypes.string.isRequired,
};

describe('Icon', () => {
    it('#className', () => {
        expect(
            shallow(<Icon icon={SQUARE} className="u-test" />),
        ).toMatchSnapshot();
    });

    it('#title', () => {
        expect(
            shallow(<Icon icon={SQUARE} title="Test title" />),
        ).toMatchSnapshot();
    });

    describe('#icon', () => {
        it('svg path', () => {
            expect(shallow(<Icon icon={SQUARE} />)).toMatchSnapshot();
        });

        it('svg path array', () => {
            expect(shallow(<Icon icon={[SQUARE, SQUARE]} />)).toMatchSnapshot();
        });

        it('inline svg ref', () => {
            expect(shallow(<Icon icon="#square" />)).toMatchSnapshot();
        });

        it('external svg ref', () => {
            expect(shallow(<Icon icon="test.svg#square" />)).toMatchSnapshot();
        });

        it('custom', () => {
            expect(
                shallow(<Icon icon={<CustomIcon icon="square" />} />),
            ).toMatchSnapshot();
        });
    });
});
