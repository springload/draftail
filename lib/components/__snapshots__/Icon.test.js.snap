// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon #className 1`] = `
<svg
  aria-hidden={true}
  aria-label={null}
  className="Draftail-Icon u-test"
  height="16"
  role={null}
  viewBox="0 0 1024 1024"
  width="16"
>
  <path
    d="M10 10 H 90 V 90 H 10 Z"
  />
</svg>
`;

exports[`Icon #icon custom 1`] = `
<CustomIcon
  icon="square"
/>
`;

exports[`Icon #icon external svg ref 1`] = `
<svg
  aria-hidden={true}
  aria-label={null}
  className="Draftail-Icon "
  height="16"
  role={null}
  viewBox="0 0 1024 1024"
  width="16"
>
  <use
    xlinkHref="test.svg#square"
  />
</svg>
`;

exports[`Icon #icon inline svg ref 1`] = `
<svg
  aria-hidden={true}
  aria-label={null}
  className="Draftail-Icon "
  height="16"
  role={null}
  viewBox="0 0 1024 1024"
  width="16"
>
  <use
    xlinkHref="#square"
  />
</svg>
`;

exports[`Icon #icon svg path 1`] = `
<svg
  aria-hidden={true}
  aria-label={null}
  className="Draftail-Icon "
  height="16"
  role={null}
  viewBox="0 0 1024 1024"
  width="16"
>
  <path
    d="M10 10 H 90 V 90 H 10 Z"
  />
</svg>
`;

exports[`Icon #icon svg path array 1`] = `
<svg
  aria-hidden={true}
  aria-label={null}
  className="Draftail-Icon "
  height="16"
  role={null}
  viewBox="0 0 1024 1024"
  width="16"
>
  <path
    d="M10 10 H 90 V 90 H 10 Z"
    key="0"
  />
  <path
    d="M10 10 H 90 V 90 H 10 Z"
    key="1"
  />
</svg>
`;

exports[`Icon #title 1`] = `
<svg
  aria-hidden={null}
  aria-label="Test title"
  className="Draftail-Icon "
  height="16"
  role="img"
  viewBox="0 0 1024 1024"
  width="16"
>
  <path
    d="M10 10 H 90 V 90 H 10 Z"
  />
</svg>
`;
